File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n
  | n < 1 = Nothing
  | aliquotSum < n = Just Deficient
  | aliquotSum == n = Just Perfect
  | otherwise = Just Abundant
  where 
  divisibleBy n m = n `mod` m == 0 
  aliquotSum = sum $ filter (divisibleBy n) [1..n `div` 2]
  

File: lua/perfect-numbers/perfect-numbers.lua
local function aliquot_sum(n)
	local sum = 0
	for i = 1, math.floor(n / 2) do
		if n % i == 0 then
			sum = sum + i
		end
	end
	return sum
end

local function classify(n)
	local aliquot = aliquot_sum(n)
	if aliquot == n then
		return "perfect"
	elseif aliquot < n then
		return "deficient"
	else
		return "abundant"
	end
end

return { aliquot_sum = aliquot_sum, classify = classify }
